namenode/namenode.go

```diff
@@ -44,27 +55,279 @@ func (server *NameNode) Create(path string, success *bool) error {
 
 func (server *NameNode) Exists(path string, exists *bool) error {
 	slog.Info("Exists request", "path", path)
-	server.globalRWLock.RLock()
-	defer server.globalRWLock.RUnlock()
+	server.inodeRWLock.RLock()
+	defer server.inodeRWLock.RUnlock()
 	_, *exists = server.inodes[path]
 	return nil
 }
 
+// TODO: 2pc, notify all the datanodes to delete the blocks before responding
 // This RPC doesn't return anything
 func (server *NameNode) Delete(path string, unused *bool) error {
 	slog.Info("Delete request", "path", path)
 
-	server.globalRWLock.Lock()
-	defer server.globalRWLock.Unlock()
+	server.inodeRWLock.Lock()
+	defer server.inodeRWLock.Unlock()
 
 	delete(server.inodes, path)
 	return nil
 }
 
 func (server *NameNode) GetAttributes(path string, fileAttributes *common.FileAttributes) error {
 	slog.Info("GetAttributes request", "path", path)
-	server.globalRWLock.RLock()
-	defer server.globalRWLock.RUnlock()
+	server.inodeRWLock.RLock()
+	defer server.inodeRWLock.RUnlock()
 	*fileAttributes = server.inodes[path].fileAttributes
 	return nil
 }
+
+func (server *NameNode) RegisterDataNode(dataNodeEndpoint string, success *bool) error {
+	slog.Info("RegisterDataNode request", "dataNodeEndpoint", dataNodeEndpoint)
+
+	server.inodeRWLock.Lock()
+	defer server.inodeRWLock.Unlock()
+
+	// TODO: Check if the datanode is already registered
+	server.datanodes[dataNodeEndpoint] = DataNodeInfo{
+		heartbeatReceived: true,
+		isAlive:           true,
+	}
+
+	*success = true
+	return nil
+}
+
+// NameNode receives block report from DataNode,
+// and updates all block storage info in the corresponding inode
+func (server *NameNode) ReportBlock(blockReport common.BlockReport, success *bool) error {
+	slog.Info("ReportBlock request", "dataNodeEndpoint", blockReport.Endpoint,
+		slog.Any("Block MetaData", blockReport.BlockMetadata))
+
+	server.inodeRWLock.Lock()
+	defer server.inodeRWLock.Unlock()
+
+	// Delete all prior records of block storage info from the datanode
+	for _, inode := range server.inodes {
+		inode.rwlock.Lock()
+		for blockID, storageInfo := range inode.storageInfo {
+			var newStorageNodes []string
+			for _, storageNode := range storageInfo.DataNodes {
```

Comment:
I still don't understand why a nested loop is needed. I may need more time.

Reply:
Now I basically understand. Since the number of replicas is small, the inner loop won't have many iterations. I feel here should definitely have some comments describing the meaning of each loop.

Reply:
Actually what I'm doing here is just to remove the reporting datanode from the storage node list. But I can't find a proper function in GoLang (like .remove() in python) to do that without iterating the list and constructing a new list. Do you have any idea?

Reply:
Yes you're right, so I decided to still use the list